src/object.c

```diff
@@ -211,6 +211,316 @@ robj *dupStringObject(const robj *o) {
+            lobj = createQuicklistObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    zfree(lobj->ptr);
+    lobj->ptr = quicklistDup(o->ptr);
+    return lobj;
+}
+/* This is a helper function for the COPY command.
+ * Duplicate a set object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupSetObject(robj *o) {
+    robj *set;
+    setTypeIterator *si;
+    sds elesds;
+    int64_t intobj;
+    int encoding;
+
+    serverAssert(o->type == OBJ_SET);
+
+    /* Create a new set object that have the same encoding as the original object's encoding */
+    switch (o->encoding) {
+        case OBJ_ENCODING_INTSET:
+            set = createIntsetObject();
+            break;
+        case OBJ_ENCODING_HT:
+            set = createSetObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    si = setTypeInitIterator(o);
+    while ((encoding = setTypeNext(si, &elesds, &intobj)) != -1) {
+        if (encoding == OBJ_ENCODING_INTSET) {
+            elesds = sdsfromlonglong(intobj);
+            setTypeAdd(set, elesds);
+            sdsfree(elesds);
+        } else {
+            setTypeAdd(set, elesds);
+        }
+    }
+    setTypeReleaseIterator(si);
+    return set;
+}
+
+/* This is a helper function for the COPY command.
+ * Duplicate a zset object, with the guarantee that the returned object
+ * has the same encoding as the original one.
+ * 
+ * The resulting object always has refcount set to 1 */
+robj *dupZsetObject(robj *o) {
+    robj *zobj;
+    long llen;
+    int retflags = ZADD_NONE;
+
+    serverAssert(o->type == OBJ_ZSET);
+    
+    /* Create a new zset object that have the same encoding as the original object's encoding */
+    switch (o->encoding) {
+        case OBJ_ENCODING_ZIPLIST:
+            zobj = createZsetZiplistObject();
+            break;
+        case OBJ_ENCODING_SKIPLIST:
+            zobj = createZsetObject();
+            break;
+        default:
+            serverPanic("Wrong encoding.");
+            break;
+    }
+    llen = zsetLength(o);
+    if (zobj->encoding == OBJ_ENCODING_ZIPLIST) {
+        unsigned char *zl = o->ptr;
+        size_t sz = ziplistBlobLen(zl);
+        unsigned char *new_zl = zmalloc(sz);
+        memcpy(new_zl, zl, sz);
+        zfree(zobj->ptr);
+        zobj->ptr = new_zl;
+    } else if (zobj->encoding == OBJ_ENCODING_SKIPLIST) {
+        zset *zs = o->ptr;
+        zskiplist *zsl = zs->zsl;
+        zskiplistNode *ln;
+        sds ele;
+
+        /* Add a score-element pair to a new zset object which encoding is SKIPLIST. */
+        ln = zsl->header->level[0].forward;
+        while (llen--) {
+            ele = ln->ele;
+            zsetAdd(zobj, ln->score, ele, &retflags, NULL);
+            ln = ln->level[0].forward;
+        }
```

Comment:
i recently remembered that zslInsert (used by zsetAdd) is more efficient when iterating on the list backwards: 
https://github.com/redis/redis/blob/f6546eff452e9868569610c797765db8f56763e3/src/rdb.c#L855-L861

it may be a topic for another PR to improved zslInsert, but maybe in the meanwhile we wanna run backwards here too (and put a comment saying why)?